### ISSUE TITLE
Standardize exports

### ISSUE THREAD

**Issue by julien** [liferay/liferay-frontend-guidelines#21]
Should we standarize one way we define our module `exports`?

We can either use `export default Thing` or `export {Thing}`;

This is probably related to #12 as well.

**Comment by wincent** [comment]
Previous discussion about this: https://github.com/liferay/clay/pull/1453

(I think it has been discussed in other places too, but that is the one I can find.)

**Comment by wincent** [comment]
Ignoring the question of whether to use `default`, one thing I'd like the recommendations to include is something compatible with [this advice](https://github.com/liferay/liferay-frontend-guidelines/blob/9a2c09739712f39a4e89f9c7b01398f4b52999b2/guidelines/general/file_names.md#how-do-i-pick-a-descriptive-camelcase-name-for-my-file-with-multiple-exports-but-no-default-export) that I put in the file-naming guidelines:

> - Prefer small files that each export one thing as opposed to a large bundle that contains many things.
> - One place where it is often appropriate to export many things is the file indicated by the "main" property in the "package.json": in this case, "index.js" is often a good name for this file.
> - When creating a file with multiple exports is the most practical choice, pick a name that describes the purpose of the contents: if you cannot identify a single purpose, that is a sign that you should find a way to split the file.

**Comment by bryceosterhaus** [comment]
Personally, I don't mind using `export default` but I I'd also be fine only using named exports.

Usually when I `export default` its the main use of the file, such as a component or a single function file. And then for named exports, I usually see them as helpers to the `default` or they are a collection of helpers in a single file. This is all pretty subjective but its generally what I've been following.

**Comment by julien** [comment]
I agree with of what both @wincent  and @bryceosterhaus  wrote, and I don't want to prohibit `export default`, but I just think we should avoid using "both" (as a habbit) if it doesn't make sense

For example, this isn't really useful.

```javascript
class Thing {
     doStuff() {

    }
}
export default Thing;
export {Thing};
```

**Comment by bryceosterhaus** [comment]
One random tidbit, if named exports generally seem to work better with ts+vscode. From experience, it seems that named exports tend to be more identifiable with vscode's auto-importing than the default exports do. Could just be my editor set up, but I generally never have issues with named exports.

**Comment by wincent** [comment]
Just wanted to add this [from here](

>  ## Default exports are favored #
>
>  The module syntax suggesting that the default export “is” the module may seem a bit strange, but it makes sense if you consider that one major design goal was to make default exports as convenient as possible. Quoting [David Herman](http://esdiscuss.org/topic/moduleimport#content-0):
>
> > ECMAScript 6 favors the single/default export style, and gives the sweetest syntax to importing the default. Importing named exports can and even should be slightly less concise.

**Comment by p2kmgcl** [comment]
Brief question, what about exports before file end? like

```js
export default function Card() {
  return `...`;
}
```

or

```js
export const SUPER_IMPORTANT = false;

class Tarta {
  // ...
}
```

(Personally I don't like them but I've seen it somewhere)

**Comment by wincent** [comment]
I think it's fine to define and export in-place; that way, as I read through a module, I know what is "public" as I'm reading and don't have to wait until the end. It's true that if the exports aren't all grouped at the end I can't jump to the end of the file to see everything at once, but if I want to know everything exported I can easily enough search for the word "export".

**Comment by julien** [comment]
@p2kmgcl  I've used the "first example" style in various places a bit like this...

```javascript

// MyComponent.es.js
const  MyComponent(props) => {
    // Que sea rica
};

export default function({context, props}) {
    <MyContext.Provider value={context}>
        <MyComponent {...props} />
    </MyContext.Provider>
}
```
